src/recovery/mod.rs

```diff
@@ -1071,6 +1073,111 @@ fn sub_abs(lhs: Duration, rhs: Duration) -> Duration {
     }
 }
 
+// We don't need to log all qlog metrics every time there is a recovery event.
+// Instead, we can log only the MetricsUpdated event data fields that we care
+// about, only when they change. To support this, the QLogMetrics structure
+// keeps a running picture of the fields.
+#[derive(Default)]
+#[cfg(feature = "qlog")]
+struct QlogMetrics {
```

Comment:
Could maybe just use the `qlog::event::Event` type rather than having a new struct?

Reply:
I tried something similar to your suggestion. The ugly thing is that the event data is `pub enum EventData` of variant `MetricsUpdated`, which made it harder to access and manipulate the actual data after the object was created. Furthermore, since qlog uses Option for every field, it makes comparison and update code a bit more onerous, especially where some fields use floats and equivalence comparison is more cumbersome.